### PR TITLE
Add map visualization - displays k-area

### PR DESCRIPTION
**Pull request by ConnectedSystems**
**Currently throws a recoverable error.**

Seems like some issue within GLMakie or GeoMakie, so marking this as a draft PR until the issue is resolved.

```julia

dom = ADRIA.load_domain("... some domain ...")

ADRIA.viz.map(dom)

# or 

# rs = ADRIA.load_results("... some results ...")
# ADRIA.viz.map(rs)
```

image

### CODE REVIEW

**Comment by ConnectedSystems** [comment]
```julia
rs = ADRIA.load_results("some result set")

tac = ADRIA.metrics.total_absolute_cover(rs)
m_tac = ADRIA.metrics.per_site(median, tac)
ADRIA.viz.map(rs, m_tac)
```

image


You can now also pass in a vector of locations to highlight, currently using a `:dashdot` line style.
The approach is a bit clunky. Open to suggestions here.

```julia
highlight = fill((:red, 0.0), 216)
highlight[1] = (:red, 1.0)
ADRIA.viz.map(rs, m_tac; opts=Dict(:highlight => highlight))
```

image


Zooming in:

image

**Comment by ConnectedSystems** [comment]
@Rosejoycrocker @Zapiano 

Could one (ideally both) of you review this today please?

File: src/metrics/site_level.jl [hunk 1]
```diff
@@ -7,20 +7,29 @@ using Statistics, OnlineStats
...
-    @inbounds for i in 1:num_sites
-        met[i] = metric(data[sites=i])
-    end
+    # Get length of timestep dimension directly
+    #   `map` erroneously extracts every single element from the NamedDimsArray
+    #   so we use `tf` to subset the dataset.
+    # Use of JuliennedArrays is in an attempt to speed up calculation of summary statistics.
+    #   We see a small but still worthwhile improvement in practice.
+    #   see: https://stackoverflow.com/a/62040897
+    tf = axes(data, :timesteps)
```
Comment (Zapiano) [inline src/metrics/site_level.jl:26]: what `tf` stands for?

Comment (ConnectedSystems) [reply src/metrics/site_level.jl:26]: timeframe

**Comment by ConnectedSystems** [comment]
@Rosejoycrocker just a small prod to get your attention

**Comment by Rosejoycrocker** [comment]
Sorry, reviewing now

File: ext/AvizExt/AvizExt.jl [hunk 2]
```diff
@@ -302,16 +306,7 @@ function ADRIA.viz.explore(rs::ResultSet)
...
-    # Make temporary copy of GeoPackage as GeoJSON
-    tmpdir = mktempdir()
-
-    local geo_fn = joinpath(tmpdir, "Aviz_$(rs.name).geojson")
-    try
-        GDF.write(geo_fn, rs.site_data; driver="geojson")
-    catch
-        GDF.write(geo_fn, rs.site_data; geom_columns=(:geom,), driver="geojson")
-    end
+    geo_fn = make_geojson_copy(rs)
```
Comment (Rosejoycrocker) [inline ext/AvizExt/AvizExt.jl:309]: So previously the user had to provide their own geojason but now this is created directly from the geopackage?

File: ext/AvizExt/viz/spatial.jl [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+function create_map!(f, geodata, data, highlight, centroids, c_label)
+    lon = first.(centroids)
+    lat = last.(centroids)
+
+    map_buffer = 0.025
+    spatial = GeoAxis(
+        f[1, 1];
+        lonlims=(minimum(lon) - map_buffer, maximum(lon) + map_buffer),
+        latlims=(minimum(lat) - map_buffer, maximum(lat) + map_buffer),
+        title="Study Area",
```
Comment (Rosejoycrocker) [inline ext/AvizExt/viz/spatial.jl:30]: Might be good to allow optional title name input?

Comment (ConnectedSystems) [reply ext/AvizExt/viz/spatial.jl:30]: Probably

File: ext/AvizExt/AvizExt.jl (hunk 2 above)
Comment (ConnectedSystems) [reply ext/AvizExt/AvizExt.jl:309]: Yes.

In the future the whole conversion to GeoJSON will be unnecessary, this is a temporary workaround for now.
There's a change in #382 which simplifies things even further.

Comment (Rosejoycrocker) [reply ext/AvizExt/AvizExt.jl:309]: Cool, thanks

**Review by Rosejoycrocker — APPROVED** [review]
Looks good to me - should I run tests or will this not have an impact on anything tested?

**Comment by ConnectedSystems** [comment]
Shouldn't impact anything, no. I haven't written tests for visualisations and not sure if I will, at least for now.

I'll add support for user defined titles/axis options before you merge.

**Comment by ConnectedSystems** [comment]
All ready for final review and merge @Rosejoycrocker

File: ext/AvizExt/viz/spatial.jl [hunk 4]
```diff
@@ -17,20 +17,24 @@ Create a spatial choropleth figure.
...
 - `c_label` : String, label to use for color bar
+- `axis_opts` : Additional options to pass to adjust Axis attributes
+  See: https://docs.makie.org/v0.19/api/index.html#Axis
 """
-function create_map!(f, geodata, data, highlight, centroids, c_label)
+function create_map!(f, geodata, data, highlight, centroids, c_label, axis_opts)
     lon = first.(centroids)
     lat = last.(centroids)
 
+    axis_opts[:title] = get(axis_opts, :title, "Study Area")
```
Comment (Rosejoycrocker) [inline ext/AvizExt/viz/spatial.jl:27]: Does this replace "Study Area" with what is in axis_opts as the title?

Comment (ConnectedSystems) [reply ext/AvizExt/viz/spatial.jl:27]: Julia documentation:
https://docs.julialang.org/en/v1/base/collections/#Base.get

Tooltip help in VS Code (displays on mouse hover):
image

Help via the repl:

```julia-repl
julia> ?

help?> get
search: get get! getpid getkey getfield getindex getglobal getproperty gethostname get_color_palette get_zero_subnormals @get_attribute Segments SegmentationFault mergewith mergewith! RegexMatch linesegments LineSegments ArgumentError linesegments! MergeSort iter_segments @generated

  get(collection, key, default)

  Return the value stored for the given key, or the given default value if no mapping for the key is present.

  │ Julia 1.7
  │
  │  For tuples and numbers, this function requires at least Julia 1.7.

  Examples
  ≡≡≡≡≡≡≡≡≡≡

  julia> d = Dict("a"=>1, "b"=>2);

  julia> get(d, "a", 3)
  1

  julia> get(d, "c", 3)
  3
```

Comment (Rosejoycrocker) [reply ext/AvizExt/viz/spatial.jl:27]: Ah okey, cool thanks. Looks good to me so I'll merge.

**Pull request merged**